components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowConsumer.java

```diff
@@ -196,11 +200,51 @@ private void sendResponse(HttpServerExchange httpExchange, Exchange camelExchang
                 IOHelper.copy(input, output, IOHelper.DEFAULT_BUFFER_SIZE, true);
             }
         } else {
-            TypeConverter tc = getEndpoint().getCamelContext().getTypeConverter();
-            ByteBuffer bodyAsByteBuffer = tc.mandatoryConvertTo(ByteBuffer.class, body);
+            ByteBuffer bodyAsByteBuffer = this.convertBodyToByteBuffer(camelExchange, httpExchange, body);
             httpExchange.getResponseSender().send(bodyAsByteBuffer);
         }
     }
+    
+    /*
+     * when an HTTP response has status code 200 
+     * but does not have content in the body change the status code to 204
+     */
+    private void changeStatusCodeToNoContent(Exchange camelExchange, HttpServerExchange httpExchange) {
+        Message answer = camelExchange.getMessage();
+        if (httpExchange.getStatusCode() == 200) {
+            answer.getHeaders().put(Exchange.HTTP_RESPONSE_CODE, 204);
+            answer.getHeaders().put(Exchange.HTTP_RESPONSE_TEXT, "No Content");
+
+            httpExchange.setStatusCode(204);
+        }
+    }
+ 
+    private ByteBuffer convertBodyToByteBuffer(Exchange camelExchange, HttpServerExchange httpExchange, Object bodyObj) throws NoTypeConversionAvailableException {
+        
+        TypeConverter tc = getEndpoint().getCamelContext().getTypeConverter();
+        ByteBuffer bodyAsByteBuffer = tc.mandatoryConvertTo(ByteBuffer.class, bodyObj);
+        
+        if (!bodyAsByteBuffer.hasRemaining()) {
+            this.changeStatusCodeToNoContent(camelExchange, httpExchange);
```

Comment:
Can you please clarify why this check was needed?

Reply:
if there isn't anything in the `ByteBuffer` we would change the status code from 200 to 204

Reply:
should we check the body if there is an exception? I was moving it to the `else` and doing the check only if there wasn't one.

Reply:
@CodeSmell I am not really sure if this is necessary.

Reply:
yeah, please see my other comment about the logic of the status code https://github.com/apache/camel/pull/3239#discussion_r334049060